### PR TITLE
Student focus fixes & enhancements

### PR DESCRIPTION
**Pull request by ridwvn22**
### Changes
Give a short summary of the changes made in this PR. 

- Added current_user.educator.id as a hidden_field in the form  and active as hidden_field
- The student focus should be the active one: used (**.sample** )to randomize students
- Updated FactoryBot factory for StudentFocus to create its focus so that it doesn't need to be done manually in tests. 

### Link to Issue 
https://app.asana.com/0/1206510424537042/1207270673214770/f

### Comments, Notes, or  Questions
N/A

### Related Screenshots 
For FE tickets specifcially, please include a screenshot of your work!
Backend code.

### Before you submit for review, Did you...

- [X] Added relevant tests (if applicable)
- [X] Linked Asana ticket to PR
- [X] Added comments, notes, or questions for reviewwer
- [] Added relevant screenshots
- [X] Ensured ticket title is clear and concise(i.e. 'Donation Form Feature' instead of 'sfg/chels/donation-form'

### CODE REVIEW

File: app/models/student_focus.rb [hunk 1]
```diff
@@ -7,4 +7,5 @@ class StudentFocus < ApplicationRecord
   validates :student_id, uniqueness: { scope: :focus_id }
   validates :focus_id, presence: true
   has_many :identifications, dependent: :destroy
+  validates :student_id, uniqueness: { scope: :focus_id }
```
Comment (edithemmings) [inline app/models/student_focus.rb:10]: Looks like this already got added back in! I see it up on line 7

File: app/views/focuses/index.html.erb [hunk 2]
```diff
@@ -6,6 +6,7 @@
   </div>
 
   <%= form_with(model: @student_focus, url: student_focuses_path(student_id: @student.id), method: :post, local: true) do |f| %>
+  <%= f.hidden_field :educator_id, value: current_user.educator.id %>
```
Comment (edithemmings) [inline app/views/focuses/index.html.erb:9]: 🎉🎉

File: app/controllers/focuses_controller.rb [hunk 3]
```diff
@@ -5,7 +5,7 @@ class FocusesController < ApplicationController
 
   def index
     @focuses = Focus.all
-    @selected_focus = @student.focuses.first
+    @selected_focus = @student.focuses.sample
```
Comment (edithemmings) [inline app/controllers/focuses_controller.rb:8]: ^should be the active focus

File: app/controllers/focuses_controller.rb [hunk 4]
```diff
@@ -29,8 +29,9 @@ def deactivate_previous_focuses
   end
 
   def build_and_save_student_focus
+    @educator_id = current_user.educator&.id
     @student_focus = @student.student_focuses.build(student_focus_params.merge(active: true,
-                                                                               educator_id: current_user.id))
+                                                                               educator_id: @educator_id))
```
[the remark this answers is not in the document]
Comment (edithemmings) [reply app/controllers/focuses_controller.rb:34]: Ooh active will be nil though unless you add a `hidden_field` for it in the form

File: test/models/student_focus_test.rb [hunk 5]
```diff
@@ -9,8 +9,7 @@ class StudentFocusTest < ActiveSupport::TestCase
     should belong_to(:focus)
   end
   test 'Create Studentfocus' do
-    focus = create(:focus)
-    studentfocus = build(:student_focus, focus:)
+    studentfocus = create(:student_focus)
```
Comment (edithemmings) [inline test/models/student_focus_test.rb:12]: 🎉🎉

File: app/controllers/focuses_controller.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@ class FocusesController < ApplicationController
 
   def index
     @focuses = Focus.all
-    @selected_focus = @student.focuses.first
+    @selected_focus = @student.student_focuses.find_by(active: true)&.focus || @student.focuses.sample
```
Comment (edithemmings) [inline app/controllers/focuses_controller.rb:8]: I think its ok if its nil if there's no active focus

File: app/controllers/focuses_controller.rb [hunk 7]
```diff
@@ -29,8 +29,9 @@ def deactivate_previous_focuses
   end
 
   def build_and_save_student_focus
-    @student_focus = @student.student_focuses.build(student_focus_params.merge(active: true,
-                                                                               educator_id: current_user.id))
+    @educator_id = current_user.educator&.id
+    @student_focus.active = true
```
Comment (edithemmings) [inline app/controllers/focuses_controller.rb:33]: Is the form saving? Looks like `@student_focus` isn't getting built anymore

File: app/controllers/focuses_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (edithemmings) [reply app/controllers/focuses_controller.rb]: ^My original comment applies again now. You don't need to merge anything in with `student_focus_params`

**Review by crestu — COMMENTED** [review]
looks good! im just not seeing where the uniqueness validation was re-added, as per the acceptance criteria.

**Comment by ridwvn22** [comment]
> looks good! im just not seeing where the uniqueness validation was re-added, as per the acceptance criteria.

-> It was duplicately added in a previous merge, so I don't need to.

**Comment by crestu** [comment]
> > looks good! im just not seeing where the uniqueness validation was re-added, as per the acceptance criteria.
> 
> -> It was duplicately added in a previous merge, so I don't need to.

ok nice, looks like its failing some tests atm but besides that i think its ready for approval on my end

File: app/controllers/focuses_controller.rb [hunk 8]
```diff
@@ -29,8 +29,8 @@ def deactivate_previous_focuses
   end
 
   def build_and_save_student_focus
-    @student_focus = @student.student_focuses.build(student_focus_params.merge(active: true,
-                                                                               educator_id: current_user.id))
+    @educator_id = current_user.educator&.id
```
Comment (edithemmings) [inline app/controllers/focuses_controller.rb:32]: ^where is this being used?

**Review by edithemmings — DISMISSED** [review]
Looks great!! Nice work!

**Review by crestu — DISMISSED** [review]
Looks good! Nice work!

File: app/views/focuses/index.html.erb [hunk 9]
```diff
@@ -1,30 +1,32 @@
...
+  <%= form_with(model: @student_focus, url: student_focuses_path(student_id: @student.id), method: :post, local: true) do |f| %>
+  <%= f.hidden_field :educator_id, value: current_user.educator.id %>
+   <%= f.hidden_field :active, value: true %>
     <ul class="min-w-1/2 list-none flex flex-row mb-6 justify-center gap-3">
       <% @focuses.first(4).each do |focus| %>
         <li class="flex-grow-0 flex-shrink-0 border border-solid text-gray-200 py-4 px-4 rounded-lg" style="width: 200px;">
-          <input type="radio" id="<%= "radio_#{focus.id}"%>" name="student_focus[focus_id]" value="<%= focus.id %>" class="mx-2 w-2 h-4 border border-solid border-gray-200 bg-black" data-focus-form-target="radio" data-action="focus-form#change" <%= 'checked' if @selected_focus && @selected_focus.focus_id == focus.id %>>
+          <input type="radio" id="<%= "radio_#{focus.id}"%>" name="student_focus[focus_id]" value="<%= focus.id %>" class="mx-2 w-2 h-4 border border-solid border-gray-200 bg-black" <%= 'checked' if @selected_focus && @selected_focus.id == focus.id %>>
```
Comment (edithemmings) [inline app/views/focuses/index.html.erb:14]: Did you mean to remove those data attributes?

File: app/views/focuses/index.html.erb [hunk 10]
```diff
@@ -1,30 +1,32 @@
...
           <label for="<%= "radio_#{focus.id}" %>" ><%= focus.name %></label>
         </li>
       <% end %>
     </ul>
     <ul class="min-w-1/2 list-none flex flex-row my-6 justify-center gap-3">
       <% @focuses.last(4).each do |focus| %>
         <li class="flex-grow-0 flex-shrink-0 border border-solid text-gray-200 py-4 px-4 rounded-lg" style="width: 200px;">
-          <input type="radio" id="<%= "radio_#{focus.id}"%>" name="student_focus[focus_id]" value="<%= focus.id %>" class="mx-2 w-2 h-4 border border-solid border-gray-200 bg-black" data-focus-form-target="radio" data-action="focus-form#change" <%= 'checked' if @selected_focus && @selected_focus.focus_id == focus.id %>>
+          <input type="radio" id="<%= "radio_#{focus.id}"%>" name="student_focus[focus_id]" value="<%= focus.id %>" class="mx-2 w-2 h-4 border border-solid border-gray-200 bg-black" <%= 'checked' if @selected_focus && @selected_focus.id == focus.id %>>
```
Comment (edithemmings) [inline app/views/focuses/index.html.erb:22]: ^here too. Maybe merge conflicts?

File: app/views/focuses/index.html.erb [hunk 11]
```diff
@@ -1,30 +1,32 @@
...
           <label for="<%= "radio_#{focus.id}" %>" ><%= focus.name %></label>
         </li>
       <% end %>
     </ul>
-    <div class="absolute bottom-0 right-0 mb-7 mr-7">
-    <%= f.submit "Continue", class: "text-white bg-green-500 px-4 py-2 rounded-lg #{'opacity-25' unless @has_focuses}", disabled: !@has_focuses, data: { focus_form_target: "submitButton" } %>
-  </div>
+    <div class= "absolute bottom-0 right-0 mb-7 mr-7">
+      <%= f.submit "Save", class: "text-white bg-green-500 px-4 py-2 rounded-lg " %>
```
Comment (edithemmings) [inline app/views/focuses/index.html.erb:28]: Here too

**Review by edithemmings — COMMENTED** [review]
Left a few comments

**Review by edithemmings — APPROVED** [review]
Looks great! Really nice job cleaning up some goofy bugs!! 🎉

**Review by crestu — APPROVED** [review]
way to go 🚀🚀 🚀

**Pull request merged**